### PR TITLE
added TimeSharedDense layer

### PR DESCRIPTION
**Pull request by charlesollion**
TimeSharedDense layer adds a dense layer useful after recurrent layers output with 'return_sequence=True'

### CODE REVIEW

File: keras/layers/core.py [hunk 1]
```diff
@@ -138,6 +138,39 @@ def output(self, train):
...
+       Tensor output dimensions:  (nb_sample, shared_dimension, output_dim)
+
+    '''
+    def __init__(self, input_dim, output_dim, init='uniform', activation='linear', weights=None):
+        self.init = initializations.get(init)
+        self.activation = activations.get(activation)
+        self.input_dim = input_dim
+        self.output_dim = output_dim
+
+        self.input = T.matrix()
```
Comment (fchollet) [inline keras/layers/core.py]: This should be tensor3 instead of matrix.

**Comment by fchollet** [comment]
Looking good! Could you also write an entry for this layer in [the doc](https://github.com/fchollet/keras/blob/master/docs/sources/layers/core.md)?

**Comment by patyork** [comment]
Is there any documentation on how Theano does dot products over Tensor3?

My concern is that it loops over each element in tensor3.shape[0] and applies the dot product across that slice..?

However, if you reshape the tensor3 -> (shape[0]*shape[1], shape[2]) matrix, then the dot product can be applied in one op (which can be a significant speed increase). You can then reshape back to tensor3 (shape[0]', shape[1]', shape[2]'). I'm curious if Theano does this reshape automatically, but I haven't had time to test.

**Comment by charlesollion** [comment]
@patyork I think you're right and this would probably correspond to a significant speed increase.
I'll make a few speed tests and probably commit something with a reshape instead of the scan.
Thanks!

**Comment by patyork** [comment]
The input/output will still need to be a tensor3, as he said above, but I think going through 2D would be beneficial.

**Comment by fchollet** [comment]
Scan can be pretty slow on GPU, so it sounds like a good idea. We should time both solutions to have a reference point in the future : )

**Comment by patyork** [comment]
Interesting. There does not seem to be any major difference between dot product using tensor3, a manual scan over a tensor3, or reshaping a tensor3 to a matrix beforehand. There are a few exceptions:
-small batch size
-batch size that is fairly large but does not over run the processing power of the GPU

I guess the graph optimizations deal with this issue. For clarity, I would say to leave it as a dot product over a tensor3.

```
# Each of the below test times are over 10 iterations of the operation.
# Done on a GTX-980
(1, 1024, 1024)
	Tensor3: 0.0681s
	Tensor3 manual scanning: 0.0392s
	Rehaping: 0.0386s
(11, 1024, 1024)
	Tensor3: 0.4517s
	Tensor3 manual scanning: 0.4304s
	Rehaping: 0.4303s
(21, 1024, 1024)
	Tensor3: 0.8770s
	Tensor3 manual scanning: 0.8792s
	Rehaping: 1.0368s
(31, 1024, 1024)
	Tensor3: 1.7536s
	Tensor3 manual scanning: 1.7514s
	Rehaping: 1.6970s
(41, 1024, 1024)
	Tensor3: 2.6048s
	Tensor3 manual scanning: 2.5756s
	Rehaping: 2.3284s
(51, 1024, 1024)
	Tensor3: 3.4429s
	Tensor3 manual scanning: 3.4430s
	Rehaping: 3.1892s
(61, 1024, 1024)
	Tensor3: 4.0966s
	Tensor3 manual scanning: 4.0908s
	Rehaping: 4.0742s
(71, 1024, 1024)
	Tensor3: 5.1168s
	Tensor3 manual scanning: 5.3673s
	Rehaping: 5.4035s
(81, 1024, 1024)
	Tensor3: 6.3346s
	Tensor3 manual scanning: 6.4000s
	Rehaping: 6.3743s
(91, 1024, 1024)
	Tensor3: 7.2622s
	Tensor3 manual scanning: 7.2488s
	Rehaping: 7.3013s
(101, 1024, 1024)
	Tensor3: 8.2690s
	Tensor3 manual scanning: 8.2125s
	Rehaping: 8.2347s
```

**Comment by fchollet** [comment]
Good to know, thanks for running the benchmark!

**Comment by patyork** [comment]
No worries.

This is actually a great merge (I believe). The reason I was pushing the AdvancedReshape layer was so that Keras would support tesnor3 inputs to the model (which are almost a requirement for a network that has Dense layers followed by a Recurrent layer, followed by more Dense layers). I believe that this new layer will allow this, although I have not tested it as a first layer.

**Pull request merged**